node/src/tasks/bsp_submit_proof.rs

```diff
@@ -246,17 +246,59 @@ where
             .saturating_mul(event.data.forest_challenges.len() as u128)
             .saturating_mul(2u32.into());
 
-        let cloned_blockchain = Arc::new(self.storage_hub_handler.blockchain.clone());
+        // Get necessary data for the retry check.
+        let cloned_sh_handler = Arc::new(self.storage_hub_handler.clone());
         let cloned_event = Arc::new(event.clone());
+        let cloned_forest_root = {
+            let fs = self
+                .storage_hub_handler
+                .forest_storage_handler
+                .get(&current_forest_key)
+                .await
+                .ok_or_else(|| anyhow!("CRITICAL❗️❗️ Failed to get forest storage."))?;
+            let root = fs.read().await.root();
+            root
+        };
 
+        // This function is a check to see if we should continue to retry the submission of the proof.
+        // If this proof submission is invalid, we should not retry it, and release the forest write lock.
         let should_retry = move || {
-            let cloned_blockchain = Arc::clone(&cloned_blockchain);
+            let cloned_sh_handler = Arc::clone(&cloned_sh_handler);
             let cloned_event = Arc::clone(&cloned_event);
+            let cloned_forest_root = Arc::new(cloned_forest_root);
 
+            // Check:
+            // - If the proof is outdated.
+            // - If the Forest root of the BSP has changed.
             Box::pin(async move {
-                Self::check_if_proof_is_outdated(&cloned_blockchain, &cloned_event)
-                    .await
-                    .is_ok()
+                let current_forest_key = CURRENT_FOREST_KEY.to_vec();
+                let is_proof_outdated =
+                    Self::check_if_proof_is_outdated(&cloned_sh_handler.blockchain, &cloned_event)
+                        .await
+                        .is_err();
+                let has_forest_root_changed = {
+                    let fs = cloned_sh_handler
+                        .forest_storage_handler
+                        .get(&current_forest_key)
+                        .await;
+
+                    match fs {
+                        Some(fs) => fs.read().await.root() != *cloned_forest_root,
+                        None => true,
```

Comment:
Here `fs` might be none because the forest key has changed after a reorg? Or why?